File: back-go/app/features/content/services/content_service.go
package services

import (
	"fmt"
	"math"
	"strings"
	"time"
	"nareshka-backend/app/shared/entities"
	"nareshka-backend/app/features/content/dto"

	"gorm.io/gorm"
)

type ContentService struct {
	db *gorm.DB
}

func NewContentService(db *gorm.DB) *ContentService {
	return &ContentService{
		db: db,
	}
}

func (s *ContentService) GetContentBlocks(req *dto.ContentBlockFilterRequest, userID *uint) (*dto.ContentBlockListResponse, error) {
	var blocks []entities.ContentBlock
	var total int64

	query := s.db.Model(&entities.ContentBlock{}).Where("is_active = ?", true)

	// Apply filters
	if req.Category != "" {
		query = query.Joins("JOIN content_files ON content_blocks.file_id = content_files.id").
			Where("content_files.main_category = ?", req.Category)
	}
	if req.SubCategory != "" {
		query = query.Joins("JOIN content_files ON content_blocks.file_id = content_files.id").
			Where("content_files.sub_category = ?", req.SubCategory)
	}
	if req.FileID != "" {
		query = query.Where("file_id = ?", req.FileID)
	}
	if req.Level != nil {
		query = query.Where("block_level = ?", *req.Level)
	}
	if req.MinOrder != nil {
		query = query.Where("order_in_file >= ?", *req.MinOrder)
	}
	if req.MaxOrder != nil {
		query = query.Where("order_in_file <= ?", *req.MaxOrder)
	}
	if req.Search != "" {
		searchTerm := "%" + strings.ToLower(req.Search) + "%"
		query = query.Where(
			"LOWER(block_title) LIKE ? OR LOWER(text_content) LIKE ? OR LOWER(code_content) LIKE ?",
			searchTerm, searchTerm, searchTerm,
		)
	}

	// Get total count
	if err := query.Count(&total).Error; err != nil {
		return nil, fmt.Errorf("failed to count content blocks: %w", err)
	}

	// Apply sorting
	sortBy := req.SortBy
	if sortBy == "" {
		sortBy = "order_in_file"
	}
	sortDir := req.SortDir
	if sortDir != "desc" {
		sortDir = "asc"
	}
	query = query.Order(fmt.Sprintf("%s %s", sortBy, sortDir))

	// Apply pagination
	offset := (req.Page - 1) * req.Limit
	query = query.Offset(offset).Limit(req.Limit)

	// Load blocks with relations
	preloadQuery := query.Preload("File")
	if userID != nil {
		preloadQuery = preloadQuery.Preload("UserProgress", "user_id = ?", *userID)
	}

	if err := preloadQuery.Find(&blocks).Error; err != nil {
		return nil, fmt.Errorf("failed to get content blocks: %w", err)
	}

	// Convert to response
	blockResponses := make([]dto.ContentBlockResponse, len(blocks))
	for i, block := range blocks {
		blockResponses[i] = dto.ContentBlockToResponse(&block)
	}

	totalPages := int(math.Ceil(float64(total) / float64(req.Limit)))

	return &dto.ContentBlockListResponse{
		Blocks:     blockResponses,
		Total:      total,
		Page:       req.Page,
		Limit:      req.Limit,
		TotalPages: totalPages,
	}, nil
}

func (s *ContentService) GetContentBlock(id string, userID *uint) (*dto.ContentBlockResponse, error) {
	var block entities.ContentBlock
	query := s.db.Preload("File")
	
	if userID != nil {
		query = query.Preload("UserProgress", "user_id = ?", *userID)
	}

	if err := query.First(&block, "id = ? AND is_active = ?", id, true).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, fmt.Errorf("content block not found")
		}
		return nil, fmt.Errorf("failed to get content block: %w", err)
	}

	response := dto.ContentBlockToResponse(&block)
	return &response, nil
}

func (s *ContentService) GetContentFiles(req *dto.ContentFileFilterRequest) (*dto.ContentFileListResponse, error) {
	var files []entities.ContentFile
	var total int64

	query := s.db.Model(&entities.ContentFile{}).Where("is_active = ?", true)

	// Apply filters
	if req.MainCategory != "" {
		query = query.Where("main_category = ?", req.MainCategory)
	}
	if req.SubCategory != "" {
		query = query.Where("sub_category = ?", req.SubCategory)
	}
	if req.Search != "" {
		searchTerm := "%" + strings.ToLower(req.Search) + "%"
		query = query.Where("LOWER(webdav_path) LIKE ?", searchTerm)
	}

	// Get total count
	if err := query.Count(&total).Error; err != nil {
		return nil, fmt.Errorf("failed to count content files: %w", err)
	}

	// Apply sorting
	sortBy := req.SortBy
	if sortBy == "" {
		sortBy = "created_at"
	}
	sortDir := req.SortDir
	if sortDir != "asc" {
		sortDir = "desc"
	}
	query = query.Order(fmt.Sprintf("%s %s", sortBy, sortDir))

	// Apply pagination
	offset := (req.Page - 1) * req.Limit
	query = query.Offset(offset).Limit(req.Limit)

	if err := query.Find(&files).Error; err != nil {
		return nil, fmt.Errorf("failed to get content files: %w", err)
	}

	// Convert to response
	fileResponses := make([]dto.ContentFileResponse, len(files))
	for i, file := range files {
		fileResponses[i] = dto.ContentFileToResponse(&file)
	}

	totalPages := int(math.Ceil(float64(total) / float64(req.Limit)))

	return &dto.ContentFileListResponse{
		Files:      fileResponses,
		Total:      total,
		Page:       req.Page,
		Limit:      req.Limit,
		TotalPages: totalPages,
	}, nil
}

func (s *ContentService) UpdateProgress(blockID string, userID uint, increment bool) (*dto.UserContentProgressResponse, error) {
	var progress entities.UserContentProgress

	// Check if block exists
	var block entities.ContentBlock
	if err := s.db.First(&block, "id = ? AND is_active = ?", blockID, true).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, fmt.Errorf("content block not found")
		}
		return nil, fmt.Errorf("failed to find content block: %w", err)
	}

	// Find or create progress record
	err := s.db.Where("user_id = ? AND content_block_id = ?", userID, blockID).First(&progress).Error
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			// Create new progress record
			progress = entities.UserContentProgress{
				UserID:         userID,
				ContentBlockID: blockID,
				SolvedCount:    0,
				IsCompleted:    false,
				LastAccessedAt: time.Now(),
			}
			if err := s.db.Create(&progress).Error; err != nil {
				return nil, fmt.Errorf("failed to create progress record: %w", err)
			}
		} else {
			return nil, fmt.Errorf("failed to find progress record: %w", err)
		}
	}

	// Update progress
	if increment {
		progress.SolvedCount++
	} else if progress.SolvedCount > 0 {
		progress.SolvedCount--
	}

	progress.LastAccessedAt = time.Now()
	progress.IsCompleted = progress.SolvedCount > 0

	if err := s.db.Save(&progress).Error; err != nil {
		return nil, fmt.Errorf("failed to update progress: %w", err)
	}

	return &dto.UserContentProgressResponse{
		ID:             progress.ID,
		UserID:         progress.UserID,
		ContentBlockID: progress.ContentBlockID,
		SolvedCount:    progress.SolvedCount,
		IsCompleted:    progress.IsCompleted,
		LastAccessedAt: progress.LastAccessedAt,
		CreatedAt:      progress.CreatedAt,
		UpdatedAt:      progress.UpdatedAt,
	}, nil
}

func (s *ContentService) GetCategories() (*dto.CategoriesResponse, error) {
	var results []struct {
		MainCategory string
		SubCategory  string
		BlockCount   int64
	}

	err := s.db.Table("content_files").
		Select("content_files.main_category, content_files.sub_category, COUNT(content_blocks.id) as block_count").
		Joins("LEFT JOIN content_blocks ON content_files.id = content_blocks.file_id AND content_blocks.is_active = true").
		Where("content_files.is_active = true").
		Group("content_files.main_category, content_files.sub_category").
		Find(&results).Error

	if err != nil {
		return nil, fmt.Errorf("failed to get categories: %w", err)
	}

	// Group by main category
	categoryMap := make(map[string]*dto.CategoryResponse)
	for _, result := range results {
		if _, exists := categoryMap[result.MainCategory]; !exists {
			categoryMap[result.MainCategory] = &dto.CategoryResponse{
				Name:          result.MainCategory,
				SubCategories: []string{},
				BlockCount:    0,
			}
		}

		category := categoryMap[result.MainCategory]
		if result.SubCategory != "" {
			// Check if subcategory already exists
			exists := false
			for _, subCat := range category.SubCategories {
				if subCat == result.SubCategory {
					exists = true
					break
				}
			}
			if !exists {
				category.SubCategories = append(category.SubCategories, result.SubCategory)
			}
		}
		category.BlockCount += result.BlockCount
	}

	// Convert map to slice
	categories := make([]dto.CategoryResponse, 0, len(categoryMap))
	for _, category := range categoryMap {
		categories = append(categories, *category)
	}

	return &dto.CategoriesResponse{
		Categories: categories,
	}, nil
}

File: back/app/features/content/services/content_service.py
"""
Content Service - бизнес-логика для работы с контентом

Содержит всю бизнес-логику для работы с образовательным контентом:
- Получение файлов и блоков контента
- Управление прогрессом пользователя
- Работа с категориями и поиском
"""

from typing import List, Optional, Tuple

from app.core.logging import get_logger
from app.features.content.dto.requests import ProgressAction
from app.features.content.dto.responses import (
    ContentBlockResponse,
    ContentBlockWithProgressResponse,
    ContentFileResponse,
)
from app.features.content.exceptions.content_exceptions import (
    InvalidProgressActionError,
)
from app.features.content.repositories.content_repository import ContentRepository
from app.shared.models.content_models import UserContentProgress as ContentBlockProgress

logger = get_logger(__name__)


class ContentService:
    """Сервис для работы с образовательным контентом"""

    def __init__(self, content_repository: ContentRepository):
        self.content_repository = content_repository

    async def get_content_files(
        self,
        page: int = 1,
        limit: int = 10,
        main_category: Optional[str] = None,
        sub_category: Optional[str] = None,
        webdav_path: Optional[str] = None,
    ) -> Tuple[List[ContentFileResponse], int]:
        """Получение файлов контента с пагинацией и фильтрацией"""
        logger.debug(
            "Getting content files",
            extra={
                "page": page,
                "limit": limit,
                "main_category": main_category,
                "sub_category": sub_category,
                "webdav_path": webdav_path,
            },
        )

        files, total = await self.content_repository.get_content_files(
            page=page,
            limit=limit,
            main_category=main_category,
            sub_category=sub_category,
            webdav_path=webdav_path,
        )

        # Конвертируем в response DTOs
        file_responses = [
            ContentFileResponse(
                id=file.id,
                createdAt=file.createdAt,
                updatedAt=file.updatedAt,
                webdavPath=file.webdavPath,
                mainCategory=file.mainCategory,
                subCategory=file.subCategory,
                lastFileHash=file.lastFileHash,
            )
            for file in files
        ]

        logger.info(f"Retrieved {len(file_responses)} content files (total: {total})")
        return file_responses, total

    async def get_content_blocks(
        self,
        page: int = 1,
        limit: int = 10,
        webdav_path: Optional[str] = None,
        main_category: Optional[str] = None,
        sub_category: Optional[str] = None,
        file_path_id: Optional[str] = None,
        search_query: Optional[str] = None,
        sort_by: str = "orderInFile",
        sort_order: str = "asc",
        user_id: Optional[int] = None,
    ) -> Tuple[List[ContentBlockResponse], int]:
        """Получение блоков контента с пагинацией, фильтрацией и прогрессом пользователя"""
        logger.debug(
            "Getting content blocks",
            extra={
                "page": page,
                "limit": limit,
                "user_id": user_id,
                "search_query": search_query,
                "sort_by": sort_by,
            },
        )

        blocks, total = await self.content_repository.get_content_blocks(
            page=page,
            limit=limit,
            webdav_path=webdav_path,
            main_category=main_category,
            sub_category=sub_category,
            file_path_id=file_path_id,
            search_query=search_query,
            sort_by=sort_by,
            sort_order=sort_order,
        )

        # Если пользователь аутентифицирован, добавляем прогресс
        if user_id:
            block_responses = []
            for block in blocks:
                progress = await self.content_repository.get_user_content_progress(
                    user_id, block.id
                )

                # Создаем response с прогрессом
                block_response = ContentBlockWithProgressResponse(
                    id=block.id,
                    createdAt=block.createdAt,
                    updatedAt=block.updatedAt,
                    fileId=block.fileId,
                    pathTitles=block.pathTitles,
                    blockTitle=block.blockTitle,
                    blockLevel=block.blockLevel,
                    orderInFile=block.orderInFile,
                    textContent=block.textContent,
                    codeContent=block.codeContent,
                    codeLanguage=block.codeLanguage,
                    isCodeFoldable=block.isCodeFoldable,
                    codeFoldTitle=block.codeFoldTitle,
                    extractedUrls=block.extractedUrls,
                    companies=block.companies,
                    rawBlockContentHash=block.rawBlockContentHash,
                    userProgress=progress.solvedCount if progress else 0,
                )

                # Добавляем информацию о файле если доступна
                if hasattr(block, "file") and block.file:
                    block_response.file = ContentFileResponse(
                        id=block.file.id,
                        createdAt=block.file.createdAt,
                        updatedAt=block.file.updatedAt,
                        webdavPath=block.file.webdavPath,
                        mainCategory=block.file.mainCategory,
                        subCategory=block.file.subCategory,
                        lastFileHash=block.file.lastFileHash,
                    )

                block_responses.append(block_response)

            logger.info(
                f"Retrieved {len(block_responses)} content blocks with progress (total: {total})"
            )
            return block_responses, total
        else:
            # Без прогресса пользователя
            block_responses = []
            for block in blocks:
                block_response = ContentBlockResponse(
                    id=block.id,
                    createdAt=block.createdAt,
                    updatedAt=block.updatedAt,
                    fileId=block.fileId,
                    pathTitles=block.pathTitles,
                    blockTitle=block.blockTitle,
                    blockLevel=block.blockLevel,
                    orderInFile=block.orderInFile,
                    textContent=block.textContent,
                    codeContent=block.codeContent,
                    codeLanguage=block.codeLanguage,
                    isCodeFoldable=block.isCodeFoldable,
                    codeFoldTitle=block.codeFoldTitle,
                    extractedUrls=block.extractedUrls,
                    companies=block.companies,
                    rawBlockContentHash=block.rawBlockContentHash,
                )

                # Добавляем информацию о файле если доступна
                if hasattr(block, "file") and block.file:
                    block_response.file = ContentFileResponse(
                        id=block.file.id,
                        createdAt=block.file.createdAt,
                        updatedAt=block.file.updatedAt,
                        webdavPath=block.file.webdavPath,
                        mainCategory=block.file.mainCategory,
                        subCategory=block.file.subCategory,
                        lastFileHash=block.file.lastFileHash,
                    )

                block_responses.append(block_response)

            logger.info(
                f"Retrieved {len(block_responses)} content blocks (total: {total})"
            )
            return block_responses, total

    async def get_content_block_by_id(
        self, block_id: str, user_id: Optional[int] = None
    ) -> ContentBlockResponse:
        """Получение конкретного блока контента по ID"""
        logger.debug(f"Getting content block by ID: {block_id} (user_id: {user_id})")

        # Получаем блок (исключение будет выброшено автоматически если не найден)
        block = await self.content_repository.get_content_block_by_id(block_id)

        if user_id:
            # Получаем прогресс пользователя
            progress = await self.content_repository.get_user_content_progress(
                user_id, block_id
            )

            response = ContentBlockWithProgressResponse(
                id=block.id,
                createdAt=block.createdAt,
                updatedAt=block.updatedAt,
                fileId=block.fileId,
                pathTitles=block.pathTitles,
                blockTitle=block.blockTitle,
                blockLevel=block.blockLevel,
                orderInFile=block.orderInFile,
                textContent=block.textContent,
                codeContent=block.codeContent,
                codeLanguage=block.codeLanguage,
                isCodeFoldable=block.isCodeFoldable,
                codeFoldTitle=block.codeFoldTitle,
                extractedUrls=block.extractedUrls,
                companies=block.companies,
                rawBlockContentHash=block.rawBlockContentHash,
                userProgress=progress.solvedCount if progress else 0,
            )
        else:
            response = ContentBlockResponse(
                id=block.id,
                createdAt=block.createdAt,
                updatedAt=block.updatedAt,
                fileId=block.fileId,
                pathTitles=block.pathTitles,
                blockTitle=block.blockTitle,
                blockLevel=block.blockLevel,
                orderInFile=block.orderInFile,
                textContent=block.textContent,
                codeContent=block.codeContent,
                codeLanguage=block.codeLanguage,
                isCodeFoldable=block.isCodeFoldable,
                codeFoldTitle=block.codeFoldTitle,
                extractedUrls=block.extractedUrls,
                companies=block.companies,
                rawBlockContentHash=block.rawBlockContentHash,
            )

        # Добавляем информацию о файле
        if hasattr(block, "file") and block.file:
            response.file = ContentFileResponse(
                id=block.file.id,
                createdAt=block.file.createdAt,
                updatedAt=block.file.updatedAt,
                webdavPath=block.file.webdavPath,
                mainCategory=block.file.mainCategory,
                subCategory=block.file.subCategory,
                lastFileHash=block.file.lastFileHash,
            )

        logger.info(f"Retrieved content block: {block.blockTitle}")
        return response

    async def get_content_categories(self) -> List[str]:
        """Получение списка всех основных категорий контента"""
        logger.debug("Getting content categories")

        categories = await self.content_repository.get_content_categories()

        logger.info(f"Retrieved {len(categories)} content categories")
        return categories

    async def get_content_subcategories(self, category: str) -> List[str]:
        """Получение списка подкатегорий для указанной категории"""
        logger.debug(f"Getting subcategories for category: {category}")

        subcategories = await self.content_repository.get_content_subcategories(
            category
        )

        logger.info(f"Retrieved {len(subcategories)} subcategories for '{category}'")
        return subcategories

    async def update_content_block_progress(
        self, block_id: str, user_id: int, action: ProgressAction
    ) -> ContentBlockProgress:
        """Обновление прогресса пользователя по блоку контента"""
        logger.debug(
            f"Updating content progress: block_id={block_id}, user_id={user_id}, action={action.action}"
        )

        # Проверяем существование блока (исключение выбросится автоматически)
        await self.content_repository.get_content_block_by_id(block_id)

        # Получаем текущий прогресс
        current_progress = await self.content_repository.get_user_content_progress(
            user_id, block_id
        )
        current_count = current_progress.solvedCount if current_progress else 0

        # Вычисляем новое значение
        if action.action == "increment":
            new_count = current_count + 1
        elif action.action == "decrement":
            new_count = max(0, current_count - 1)
        else:
            raise InvalidProgressActionError(action.action)

        # Сохраняем новый прогресс
        progress = await self.content_repository.create_or_update_user_progress(
            user_id=user_id, block_id=block_id, solved_count=new_count
        )

        logger.info(
            f"Updated content progress: {current_count} -> {new_count}",
            extra={"user_id": user_id, "block_id": block_id, "action": action.action},
        )

        return progress

    async def get_user_total_solved_count(self, user_id: int) -> int:
        """Получение общего количества решенных блоков пользователем"""
        logger.debug(f"Getting total solved count for user: {user_id}")

        count = await self.content_repository.get_user_total_solved_count(user_id)

        logger.info(f"User {user_id} has solved {count} content blocks")
        return count

    @staticmethod
    def unescape_text_content(text: Optional[str]) -> Optional[str]:
        """Утилитарная функция для обработки экранированных символов в тексте"""
        if not text:
            return text
        return text.replace("\\n", "\n").replace("\\t", "\t")
